contrib/mesos/pkg/queue/delay/delay_queue.go

```diff
@@ -0,0 +1,172 @@
+)
+
+// Queue is a thread-safe, time-based queue.
+//
+// Items can only be popped after their event time has been reached.
+// Use Add to specify a delay duration and have the event time calculated.
+// Use Offer to specify a specific event time.
+type Queue struct {
+	*priority.Queue
+	lock  *sync.RWMutex
+	cond  *sync.Cond
+	clock clock.Clock
+}
+
+func NewDelayQueue(clock clock.Clock) *Queue {
+	var lock sync.RWMutex
+	return &Queue{
+		Queue: priority.NewPriorityQueue(),
+		lock:  &lock,
+		cond:  sync.NewCond(&lock),
+		clock: clock,
+	}
+}
+
+func (q *Queue) Add(d Delayed) {
+	p := NewDelayedPriority(d, q.clock)
+
+	item := &addItem{
+		Item: priority.NewItem(d, p),
+		lock: q.lock,
+		cond: q.cond,
+	}
+	item.Push(q.Queue)
+}
+
+// addItem adds locking and broadcasting to priority.Item.Push
+type addItem struct {
+	priority.Item
+	lock *sync.RWMutex
+	cond *sync.Cond
+}
+
+func (di *addItem) Push(queue heap.Interface) {
+	di.lock.Lock()
+	defer di.lock.Unlock()
+	heap.Push(queue, di)
+	di.cond.Broadcast()
+}
+
+// If there's a eventTime reported by d.EventTime() then `d` is added to the
+// queue and this func returns true.
+func (q *Queue) Offer(d Scheduled) bool {
+	eventTime, ok := NewScheduledPriority(d)
+	if ok {
+		q.lock.Lock()
+		defer q.lock.Unlock()
+		heap.Push(q.Queue, &offerItem{
+			Item: priority.NewItem(d, eventTime),
+		})
+		q.cond.Broadcast()
+	}
+	return ok
+}
+
+// offerItem offers the value priority.Item.Push
+type offerItem struct {
+	priority.Item
+}
+
+func (di *offerItem) Push(queue heap.Interface) {
+	dq := queue.(*Queue)
+	dq.Offer(di.Value().(Scheduled))
+}
+
+// wait for the delay of the next item in the queue to expire, blocking if
+// there are no items in the queue. does not guarantee first-come-first-serve
+// ordering with respect to clients.
+func (q *Queue) Pop() interface{} {
+	// doesn't implement cancellation, will always return a non-nil value
+	return q.pop(func() pushableItem {
+		q.lock.Lock()
+		defer q.lock.Unlock()
+		for q.Len() == 0 {
+			q.cond.Wait()
+		}
+		x := heap.Pop(q.Queue)
+		item := x.(pushableItem)
+		return item
+	}, nil)
+}
+
+// returns a non-nil value from the queue, or else nil if/when cancelled; if cancel
+// is nil then cancellation is disabled and this func must return a non-nil value.
+func (q *Queue) pop(next func() pushableItem, cancel <-chan struct{}) interface{} {
+	var condCh chan struct{}
+	var delayTimer clock.Timer
+	for {
+		item := next()
+		if item == nil {
+			// cancelled
+			return nil
+		}
+		x := item.Value()
+		delayPriority := item.Priority().(Priority)
+		delayedTime := delayPriority.eventTime
+		if delayedTime.After(q.clock.Now()) {
+			// listen for calls to Add() while we're waiting for the eventTime
+			if condCh == nil {
+				condCh = make(chan struct{}, 1)
+			}
+			if delayTimer == nil {
+				delayTimer = q.clock.NewTimer(delayedTime.Sub(q.clock.Now()))
+				defer delayTimer.Stop()
+				//TODO: delayTimer.Stop() sooner
```

Comment:
I added the stop myself. It was using a leaky time.After before. I wasn't sure if I needed to stop it sooner. but the defer is probably sufficient.